nixos/modules/services/matrix/matrix-synapse.nix

```diff
@@ -0,0 +1,768 @@
+          NixOps is in use.
+        '';
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    assertions = [
+      { assertion = hasLocalPostgresDB -> config.services.postgresql.enable;
+        message = ''
+          Cannot deploy matrix-synapse with a configuration for a local postgresql database
+            and a missing postgresql service. Since 20.03 it's mandatory to manually configure the
+            database (please read the thread in https://github.com/NixOS/nixpkgs/pull/80447 for
+            further reference).
+
+            If you
+            - try to deploy a fresh synapse, you need to configure the database yourself. An example
+              for this can be found in <nixpkgs/nixos/tests/matrix-synapse.nix>
+            - update your existing matrix-synapse instance, you simply need to add `services.postgresql.enable = true`
+              to your configuration.
+
+          For further information about this update, please read the release-notes of 20.03 carefully.
+        '';
+      }
+    ];
+
+    users.users.matrix-synapse = {
+      group = "matrix-synapse";
+      home = cfg.dataDir;
+      createHome = true;
+      shell = "${pkgs.bash}/bin/bash";
+      uid = config.ids.uids.matrix-synapse;
+    };
+
+    users.groups.matrix-synapse = {
+      gid = config.ids.gids.matrix-synapse;
+    };
+
+    systemd.services.matrix-synapse = {
+      description = "Synapse Matrix homeserver";
+      after = [ "network.target" ] ++ optional hasLocalPostgresDB "postgresql.service";
+      wantedBy = [ "multi-user.target" ];
+      preStart = ''
+        ${cfg.package}/bin/synapse_homeserver \
+          --config-path ${configFile} \
+          --keys-directory ${cfg.dataDir} \
+          --generate-keys
+      '';
+      environment = {
+        PYTHONPATH = makeSearchPathOutput "lib" cfg.package.python.sitePackages [ pluginsEnv ];
+      } // optionalAttrs (cfg.withJemalloc) {
+        LD_PRELOAD = "${pkgs.jemalloc}/lib/libjemalloc.so";
+      };
+      serviceConfig = {
+        Type = "notify";
+        User = "matrix-synapse";
+        Group = "matrix-synapse";
+        WorkingDirectory = cfg.dataDir;
+        ExecStartPre = [ ("+" + (pkgs.writeShellScript "matrix-synapse-fix-permissions" ''
+          chown matrix-synapse:matrix-synapse ${cfg.dataDir}/homeserver.signing.key
+          chmod 0600 ${cfg.dataDir}/homeserver.signing.key
+        '')) ];
+        ExecStart = ''
+          ${cfg.package}/bin/synapse_homeserver \
+            ${ concatMapStringsSep "\n  " (x: "--config-path ${x} \\") ([ configFile ] ++ cfg.extraConfigFiles) }
+            --keys-directory ${cfg.dataDir}
+        '';
+        ExecReload = "${pkgs.util-linux}/bin/kill -HUP $MAINPID";
+        Restart = "on-failure";
+        UMask = "0077";
+      };
+    };
+
+    environment.systemPackages = [ registerNewMatrixUser ];
+  };
+
+  meta = {
+    buildDocsInSandbox = false;
```

Comment:
Not as long as we reference the sample_config.yaml URL for the specific package version in the settings options description.

Reply:
Not sure if I'm missing something, but can't we fix that somehow? %)

Reply:
We can link to the example documentation for synapse master.

Reply:
Actually we have multiple places now where that happens and I think its better to keep them.

https://github.com/mweinelt/nixpkgs/blob/cb07d9b0f866762370c93b886509042ef177503f/nixos/modules/services/matrix/matrix-synapse.nix#L496-L501

Reply:
when we were working on the doc split we found quite a few modules that use the version of some package to generate links in their docs. the "best" way we've found of getting those links without having the *entire* module docs built outside the sandbox was to import a module with just the dependent options and `buildDocsInSandbox = false`, but keep most of the options in a sandbox-built module. we haven't done that because it's a rather ugly change, but a full module rewrite is in a good position to do that kind of thing. (especially since synapse is growing to a size where it makes sense)

Reply:
@pennae Has this already been done before in nixpkgs? Also, if you suggest splitting out the module into a second file, that would be really ugly. Are there options to do what you said but keep it in one file?

The first naive approach would be to have both modules in a `let` binding on a single file and then merge them at the bottom. The second idea that comes to my head are submodules—can they have their own `meta` attribute that only applies to the options within them?

As a third option, and this one requires changes to the system: how about having the option to disable the sandbox selectively for a list of options in the `meta` section?

Reply:
it hasn't been done, partly because it does mess with the readability of modules and partly because we just forgot. and yes, the suggestion is to split the module into two files: one importing the other, but crucially that *other* file only contains option descriptions that can't be sandboxed. at that point the main module *can* be sandboxed since the option/doc merging will pick up the descriptions from the non-sandboxed bit and everything else from the sandboxed bit. this is only really worth it for large modules though.

Reply:
I'd argue that it isn't even worth splitting large modules into multiple files. Therefore, I'd prefer to keep the module as is pending a better support for this use case.

Reply:
> it does mess with the readability of modules

*Personally* I consider that a deal breaker. While end user experience is very important, the maintainability of a module over a large set of NixOS developers is also very important.

/opinion

Reply:
another important tidbit: the vast majority of modules *is* sandboxed. we're only in any kind of trouble if we start adding dependent strings to a lot more of them.